src/Internal/Promise.php

```diff
@@ -0,0 +1,149 @@
+<?php
+
+namespace Http\Adapter\Artax\Internal;
+
+use Amp\Promise as AmpPromise;
+use Http\Client\Exception;
+use Http\Promise\Promise as HttpPromise;
+use Psr\Http\Message\ResponseInterface;
+
+/**
+ * Promise adapter between artax and php-http, which allow to use the sendAsyncRequest and also the coroutine system by still respecting the Amp Promise.
+ *
+ * @internal
+ */
+class Promise implements HttpPromise, AmpPromise
+{
+    /** @var string */
+    private $state = HttpPromise::PENDING;
+
+    /** @var ResponseInterface */
+    private $response;
+
+    /** @var Exception */
+    private $exception;
+
+    /** @var AmpPromise */
+    private $promise;
+
+    /** @var callable|null */
+    private $onFulfilled;
+
+    /** @var callable|null */
+    private $onRejected;
+
+    /**
+     * @param AmpPromise $promise Underlying amp promise which MUST resolve with a ResponseInterface or MUST fail with a Http\Client\Exception
+     */
+    public function __construct(AmpPromise $promise)
+    {
+        $this->promise = $promise;
+        $this->promise->onResolve(function ($error, $result) {
+            if ($error !== null) {
+                $this->reject($error);
+            } else {
+                $this->resolve($result);
+            }
+        });
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function then(callable $onFulfilled = null, callable $onRejected = null)
+    {
+        $deferred = new \Amp\Deferred();
+        $newPromise = new self($deferred->promise());
+
+        $onFulfilled = $onFulfilled ?? function (ResponseInterface $response) {
+            return $response;
+        };
+
+        $onRejected = $onRejected ?? function (Exception $exception) {
+            throw $exception;
+        };
+
+        $this->onFulfilled = function (ResponseInterface $response) use ($onFulfilled, $deferred) {
+            try {
+                $deferred->resolve($onFulfilled($response));
+            } catch (Exception $exception) {
+                $deferred->fail($exception);
+            }
+        };
+
+        $this->onRejected = function (Exception $exception) use ($onRejected, $deferred) {
+            try {
+                $deferred->resolve($onRejected($exception));
+            } catch (Exception $exception) {
```

Comment:
This should catch `Throwable`.

Reply:
what does this will do if there is an error not catched here ?

Because, we mainly work with 'ResponseInterface' and 'Http\Client\Exception' in all php-http libraries, having this ensure that on plugin this is always an `Http\Client\Exception' implementation that is raised on the error side

Supplying another error in the chain will break our specification and other class (mainly plugin) that works by assuming the spec is valid.

Reply:
The adapter might be used for other cases or internal errors might happen that should be handled correctly.

Reply:
I encapsulated \Throwable into a TransferException (but maybe another exception would be better in that case ?)